### PR TITLE
TW 372: add CLI installation to CLI ref index

### PR DESCRIPTION
**Pull request by DJSanti**
## What does this PR do?

This PR compiles a more comprehensive CLI index page from the CLI repo's README.

This README contains instructions for installation and configuration, and other important information.

### CODE REVIEW

File: docs-src/cli/server/start-dev.md [hunk 1]
```diff
@@ -7,9 +7,6 @@ tags:
 	- cli
 ---
 
-The `temporal server start-dev` command starts the Temporal Server on `localhost:7233`.
```
Comment (flossypurse) [inline docs-src/cli/server/start-dev.md:10]: uh oh why is this removed?

File: docs-src/cli/README.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+```
+
+At this point you should have a server running on `localhost:7233` and a web interface at <http://localhost:8233>.
+A default Namespace has also been created.
+
+Run individual commands to interact with the local Temporal server.
+
+```bash
+temporal operator namespace list
+temporal workflow list
```
[the remark this answers is not in the document]
Comment (DJSanti) [reply docs-src/cli/README.md:66]: I was going to add it here, but it's already stated below in configuration (albeit in different words)

File: docs-src/cli/server/start-dev.md (hunk 1 above)
Comment (DJSanti) [reply docs-src/cli/server/start-dev.md:10]: It shouldn't have been. Let me go add that back.

Comment (DJSanti) [reply docs-src/cli/server/start-dev.md:10]: So it's showing up in my branch on VSCode but not on GitHub?

**Comment by DJSanti** [comment]
Table with CLI environmental variables is now a go.

**Comment by DJSanti** [comment]
So we can knock out TW 373 with this PR as well.

File: docs-src/cli/README.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
+---
+id: readme
+title: temporal readme
+sidebar_label: Install CLI
+description: How to install and configure Temporal CLI.
+tags:
+    - cli
+---
+
+[![Go Reference](https://pkg.go.dev/badge/github.com/temporalio/cli.svg)](https://pkg.go.dev/github.com/temporalio/cli)
```
Comment (bergundy) [inline docs-src/cli/README.md:10]: Why do we need the Go reference? This isn't a library.

File: docs-src/cli/README.md [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+[![ci](https://github.com/temporalio/cli/actions/workflows/ci.yml/badge.svg)](https://github.com/temporalio/cli/actions/workflows/ci.yml)
```
Comment (bergundy) [inline docs-src/cli/README.md:11]: Probably not worth putting this in the docs.

File: docs-src/cli/README.md [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+> ⚠️ Temporal CLI's API is still subject to change. ⚠️
```
Comment (bergundy) [inline docs-src/cli/README.md:13]: We need to finalize this ASAP as we're targeting a stable release in a few days.

File: docs-src/cli/README.md [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+The Temporal CLI is a tool for interacting with Temporal from the command line.
+Temporal CLI functions as a distribution of the Temporal Server and Web UI that runs as a single process with zero runtime dependencies.
+The tool supports persistence to disk and in-memory mode through SQLite.
+
+## Installation
```
Comment (bergundy) [inline docs-src/cli/README.md:19]: Note that there's an entire page in the docs that explains how to install the CLI, you might want to inline or link it.

File: docs-src/cli/README.md [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+
+```bash
+temporal server start-dev --help
+```
+
+Configure the environment with `env` commands:
+
+```bash
+temporal env set [environment options]
+```
```
Comment (bergundy) [inline docs-src/cli/README.md:79]: This requires way more explanation.

Comment (bergundy) [reply docs-src/cli/README.md:79]: It's also an advanced feature so I suggest we explain it later on.

File: docs-src/cli/README.md [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+### Enable or disable Temporal UI
+
+By default, the Temporal UI is started with Temporal CLI. The UI can be disabled via a runtime flag:
+
+```bash
+temporal server start-dev --headless
+```
+
+To build without static UI assets, use the `headless` build tag when running `go build`.
```
Comment (bergundy) [inline docs-src/cli/README.md:114]: Don't think we want to mention this here.

File: docs-src/cli/README.md [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+### Dynamic configuration
+
+Advanced configuration of the Temporal CLI requires the use of a dynamic configuration file.
+This file is created outside of the Temporal CLI; it is usually located with the service's config files.
+
+Dynamic configuration values can also be set via `--dynamic-config-value KEY=JSON_VALUE`.
+For example, to disable the search attribute cache, run:
+
+```bash
+temporal server start-dev --dynamic-config-value system.forceSearchAttributesCacheRefreshOnRead=true
```
Comment (bergundy) [inline docs-src/cli/README.md:127]: This is actually already set by default, you might want to use a different dynamic config in the example.

File: docs-src/cli/README.md [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+```
+
+Verify that `bash-completion` is installed by running `type _init_completion`.
+
+Enable completion for Temporal by adding the following code to your bash file:
+
+```
+echo 'source <(temporal completion bash)' >> ~/.bashrc
+source ~/.bashrc
+```
```
Comment (bergundy) [inline docs-src/cli/README.md:206]: Don't think we need to teach users how to install bash completion, link to the explanation is enough.

File: docs-src/cli/README.md [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+
+```bash
+go build -o dist/temporal-docgen ./cmd/temporal-docgen
+```
+
+To run all tests:
+
+```bash
+go test ./...
+```
```
Comment (bergundy) [inline docs-src/cli/README.md:228]: Let's omit this from our docs site, it's too much information.

File: docs-src/cli/README.md [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+
+## Known Issues
+
+- When consuming Temporal as a library in go mod, you may want to replace grpc-gateway with a fork to address URL escaping issue in UI. See <https://github.com/temporalio/temporalite/pull/118>
+
+- When running the executables from the Releases page in macOS you will want to allowlist `temporal` binary in `Security & Privacy` settings:
```
Comment (bergundy) [inline docs-src/cli/README.md:234]: Probably don't need the known issues here either.
We don't support consuming the CLI as a library and we don't direct users to install the binary from our releases.
@feedmeapples is this still an issue BTW? Do we need to sign the darwin release somehow?

File: docs/cli/index.md [hunk 13]
```diff
@@ -32,3 +32,226 @@ Workflow, show Workflow History, and Signal Workflow).
...
+Verify that `bash-completion` is installed by running `type _init_completion`.
+
+Enable completion for Temporal by adding the following code to your bash file:
+
+```
+echo 'source <(temporal completion bash)' >> ~/.bashrc
+source ~/.bashrc
+```
+
+## Development
```
Comment (feedmeapples) [inline docs/cli/index.md:225]: this is for CLI development and unnecessary here, can be confusing for external users

File: docs-src/cli/cmd-options/reapply-type.md [hunk 14]
```diff
@@ -7,4 +7,4 @@ tags:
 	- cli
 ---
 
-Event types to reapply after the reset point: , Signal, None. (default: All)
+Event types to reapply after the reset point: Signal, None, . (default: All)
```
Comment (flossypurse) [inline docs-src/cli/cmd-options/reapply-type.md:10]:
```suggestion
Event types to reapply after the reset point: All, Signal, None.
Default is All.
```
Probably have to fix this in source.

File: docs-src/concepts/what-is-the-temporal-cli.md [hunk 15]
```diff
@@ -11,12 +11,10 @@ tags:
...
 This tool was formerly known as `tctl next`.
 
-The Temporal CLI is a command-line tool that you can use to interact with a Temporal Cluster.
-It can perform [Namespace](/concepts/what-is-a-namespace) operations (such as register, update, and describe) and [Workflow](/concepts/what-is-a-workflow) operations (such as start
-Workflow, show Workflow History, and Signal Workflow).
+---
 
-<!-- - [How to install cli]()
-- [Environment variables for tctl](/tctl-next/environment-variables) -->
+The Temporal CLI is a command-line tool that serves as a distribution of the [Temporal Server](/concepts/what-is-the-temporal-server) and [Temporal Web UI](/concepts/what-is-the-temporal-web-ui).
```
Comment (flossypurse) [inline docs-src/concepts/what-is-the-temporal-cli.md:16]:
```suggestion
The Temporal CLI is a command-line tool that includes as a distribution of a Temporal Cluster ([Temporal Server](/concepts/what-is-the-temporal-server) , persistence (SQLite) and the [Temporal Web UI](/concepts/what-is-the-temporal-web-ui)).
```

File: docs-src/concepts/what-is-the-temporal-cli.md [hunk 16]
```diff
@@ -28,3 +26,187 @@ Workflow, show Workflow History, and Signal Workflow).
...
+### CDN
+
+To install the Temporal CLI from CDN:
+
+1. Select the binary with the corresponding platform and architecture of your system.
+2. Download the binary.
+3. Extract the downloaded archive.
+4. Add the `temporal` binary to your system PATH.
+
+## Starting the Temporal Server
```
Comment (flossypurse) [inline docs-src/concepts/what-is-the-temporal-cli.md:61]:
```suggestion
## Starting a local dev Cluster
```

You will want to update this in source obviously :)

Comment (flossypurse) [reply docs-src/concepts/what-is-the-temporal-cli.md:61]: And I know that using Cluster will be controversial to @bergundy - but embracing consistency with our terms is pretty important - and this consistency needs to apply to the CLI repo.
Since the current owners of that repo are failing to commit even after disagreeing... then we need to update this for them.

File: docs-src/concepts/what-is-the-temporal-cli.md [hunk 17]
```diff
@@ -28,3 +26,187 @@ Workflow, show Workflow History, and Signal Workflow).
...
+| `TEMPORAL_CLI_NAMESPACE`                     | Identifies a Namespace in the Temporal Workflow. Default value: "default".        | --namespace                     |
+| `TEMPORAL_CLI_TLS_CA`                        | Path to server CA certificate.                                                    | --tls-ca-path                   |
+| `TEMPORAL_CLI_TLS_CERT`                      | Path to x509 certificate.                                                         | --tls-cert-path                 |
+| `TEMPORAL_CLI_TLS_DISABLE_HOST_VERIFICATION` | Disables TLS host name verification if already enabled. Default value: false.     | --tls-disable-host-verification |
+| `TEMPORAL_CLI_TLS_KEY`                       | Path to private certificate key.                                                  | --tls-key-path                  |
+| `TEMPORAL_CLI_TLS_SERVER_NAME`               | Provides an override for the target TLS server name.                              | --tls-server-name               |
+
+### Namespace registration
```
Comment (flossypurse) [inline docs-src/concepts/what-is-the-temporal-cli.md:106]: Can we not now link to the CLI reference for this?

I don't think the README should break down every high level command usage.

This is the stuff that needs to go into the reference itself.

Having the list of "available commands" near the top will suffice.

File: docs-src/concepts/what-is-the-temporal-cli.md [hunk 18]
```diff
@@ -28,3 +26,187 @@ Workflow, show Workflow History, and Signal Workflow).
...
+
+## Configuration
+
+Use the help flag to see a full list of CLI options:
+
+```bash
+temporal server start-dev --help
+```
+
+### Environmental variables
```
Comment (flossypurse) [inline docs-src/concepts/what-is-the-temporal-cli.md:88]: Environment variables apply to the whole CLI - not just the `server start-dev` command.

Reorg the content appropriately - this is sitting right between two `server` command info bits.

**Review by flossypurse — COMMENTED** [review]
We can remove this line now 

Screenshot 2023-03-21 at 3 22 09 PM

File: docs-src/concepts/what-is-the-temporal-cli.md (hunk 17 above)
Comment (flossypurse) [reply docs-src/concepts/what-is-the-temporal-cli.md:106]: Actually - this makes sense in the context of running the dev Cluster.

File: docs-src/concepts/what-is-the-temporal-cli.md (hunk 16 above)
Comment (bergundy) [reply docs-src/concepts/what-is-the-temporal-cli.md:61]: You'll want to update the server repo too:
https://github.com/temporalio/temporal#download-and-start-temporal-server-locally
And docker compose:
https://github.com/temporalio/docker-compose#temporal-server-docker-compose-files
And pretty much every other repository that mentions Temporal server.

File: docs-src/concepts/what-is-the-temporal-cli.md [hunk 19]
```diff
@@ -28,3 +26,187 @@ Workflow, show Workflow History, and Signal Workflow).
 - [temporal server](/cli/server)
 - [temporal task-queue](/cli/task-queue/)
 - [temporal workflow](/cli/workflow/)
+
+## Installation
```
Comment (bergundy) [inline docs-src/concepts/what-is-the-temporal-cli.md:30]: Please remove the installation instructions if you're linking to the reviewed guide.

File: docs-src/concepts/what-is-the-temporal-cli.md [hunk 20]
```diff
@@ -28,3 +26,187 @@ Workflow, show Workflow History, and Signal Workflow).
...
+temporal server start-dev --headless
+```
+
+### Dynamic configuration
+
+Advanced configuration of the Temporal CLI requires the use of a dynamic configuration file.
+This file is created outside of the Temporal CLI; it is usually located with the service's config files.
+
+Dynamic configuration values can also be set via `--dynamic-config-value KEY=JSON_VALUE`.
+For example, to enable the search attribute cache, run:
```
Comment (bergundy) [inline docs-src/concepts/what-is-the-temporal-cli.md:145]:
```suggestion
For example, to enable the search attribute cache (disabled by default), run:
```

File: docs-src/concepts/what-is-the-temporal-cli.md [hunk 21]
```diff
@@ -28,3 +26,187 @@ Workflow, show Workflow History, and Signal Workflow).
...
+
+Finally, [enable completion for Temporal in your script](#auto-completion).
+
+#### macOS installation
+
+Install `bash-completion` through Homebrew:
+`brew install bash-completion@2`
+
+Add the provided code snippet to your `~/.bashrc` file:
+
```
Comment (bergundy) [inline docs-src/concepts/what-is-the-temporal-cli.md:205]: Link to the bash completion installation guide here instead of providing our own instructions.

Comment (bergundy) [reply docs-src/concepts/what-is-the-temporal-cli.md:205]: Can add this in addition to the GH link: https://repology.org/project/bash-completion/versions

File: docs/cli/index.md
Comment (bergundy) [inline docs/cli/index.md:1]: All comments from the concept page apply here as well.

**Comment by bergundy** [comment]
I have another minor comment on bash completion but otherwise LGTM, thanks!

File: docs-src/concepts/what-is-the-temporal-cli.md [hunk 22]
```diff
@@ -28,3 +21,135 @@ Workflow, show Workflow History, and Signal Workflow).
...
+`source ~/.zshrc`.
+
+### Bash auto-completion
+
+Bash auto-completion relies on `bash-completion`.
+
+Install the software with the steps provided in the [bash-completion README](https://github.com/scop/bash-completion#installation), or use your preferred package manager on your operating system.
+
+To view different bash versions, go [here](https://repology.org/project/bash-completion/versions).
```
Comment (bergundy) [inline docs-src/concepts/what-is-the-temporal-cli.md:155]: Should show how to install the completion like we do with zsh completion.


> To view different bash versions, go [here](https://repology.org/project/bash-completion/versions).

These aren't different bash versions.

**Review by djmagee — APPROVED** [review]

**Review by flossypurse — APPROVED** [review]

File: docs-src/concepts/what-is-the-temporal-cli.md [hunk 23]
```diff
@@ -28,3 +21,135 @@ Workflow, show Workflow History, and Signal Workflow).
 - [temporal server](/cli/server)
 - [temporal task-queue](/cli/task-queue/)
 - [temporal workflow](/cli/workflow/)
+
+## Installation
+
+Temporal CLI can be installed through several different methods. While most of them can be used across all operating systems, please note that Homebrew is macOS-exclusive.
```
Comment (bergundy) [inline docs-src/concepts/what-is-the-temporal-cli.md:27]: @DJSanti the note about macos is irrelevant in this context.

**Review by djmagee — COMMENTED** [review]
These comments are now moot, but because I was reviewing when the PR was merged, I'm gonna drop 'em here anyway for future reference.

File: docs-src/concepts/what-is-the-temporal-cli.md [hunk 24]
```diff
@@ -28,3 +21,135 @@ Workflow, show Workflow History, and Signal Workflow).
...
+Run individual commands to interact with the local Temporal Server.
+
+```bash
+temporal operator namespace list
+temporal workflow list
+```
+
+## Configuration
+
+Use the help flag to see a full list of CLI options:
```
Comment (bergundy) [inline docs-src/concepts/what-is-the-temporal-cli.md:52]: From the installation guide:

> The server's startup configuration can be customized using command line options. For a full list of options, run:

Comment (bergundy) [reply docs-src/concepts/what-is-the-temporal-cli.md:52]: I'd also move this section closer to the `Starting the Temporal Server` header.

**Review by bergundy — COMMENTED** [review]
@DJSanti I had a few more comments if you'd like to address in a followup PR.
Thanks!
